framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,279 @@
+
+static const std::chrono::seconds WAIT_HEADER_TIMEOUT = std::chrono::seconds(3);
+static const std::chrono::seconds WAIT_DATA_TIMEOUT = std::chrono::seconds(3);
+
+HttpInputDataSource::HttpInputDataSource(const std::string &url)
+	: InputDataSource(), mUrl(url)
+{
+	medvdbg("url: %s\n", mUrl.c_str());
+}
+
+HttpInputDataSource::HttpInputDataSource(const HttpInputDataSource &source)
+	: InputDataSource(source), mUrl(source.mUrl)
+{
+}
+
+HttpInputDataSource &HttpInputDataSource::operator=(const HttpInputDataSource &source)
+{
+	InputDataSource::operator=(source);
+	return *this;
+}
+
+bool HttpInputDataSource::open()
+{
+	medvdbg("HttpInputDataSource::open!\n");
+
+	if (isPrepare()) {
+		medvdbg("HttpInputDataSource is already opened!\n");
+		return true;
+	}
+
+	if (mStreamBuffer == nullptr) {
+		mStreamBuffer = StreamBuffer::Builder()
+								.setBufferSize(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE)
+								.setThreshold(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD)
+								.build();
+
+		if (mStreamBuffer == nullptr) {
+			meddbg("mStreamBuffer is nullptr!\n");
+			return false;
+		}
+
+		mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+		mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+	}
+
+	if (mHttpStream == nullptr) {
+		mHttpStream = HttpStream::create();
+		if (mHttpStream == nullptr) {
+			meddbg("mHttpStream is nullptr!\n");
+			return false;
+		}
+	}
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	mIsHeaderReceived = false;
+	mIsDataReceived = false;
+
+	mThread = std::thread([this]() {
+		medvdbg("download thread enter!\n");
+		//mHttpStream->addHeader("Icy-MetaData:1"); // not support now
+		mHttpStream->setHeaderCallback(HeaderCallback, (void *)this);
+		mHttpStream->setWriteCallback(WriteCallback, (void *)this);
+		if (!mHttpStream->download(mUrl)) {
+			medwdbg("download failed or terminated!\n");
+			// TODO: send network error code to upper layer later
+		}
+
+		mBufferWriter->setEndOfStream();
+		medvdbg("download thread exit!\n");
+	});
+
+	// wait for Content-Type header
+	if (!mCondv.wait_for(lock, WAIT_HEADER_TIMEOUT, [=]{ return mIsHeaderReceived; })) {
+		medwdbg("download:: wait header timeout!\n");
+		mBufferWriter->setEndOfStream();
+		return false;
+	}
+
+	audio_type_t audioType = AUDIO_TYPE_UNKNOWN;
+	if ((mContentType.find(AAC_MIME_TYPE) != std::string::npos) ||
+		(mContentType.find(AACP_MIME_TYPE) != std::string::npos)) {
+		audioType = AUDIO_TYPE_AAC;
+	} else if (mContentType.find(MPEG_MIME_TYPE) != std::string::npos) {
+		audioType = AUDIO_TYPE_MP3;
+	} else if (mContentType.find(OPUS_MIME_TYPE) != std::string::npos) {
+		audioType = AUDIO_TYPE_OPUS;
+	} else {
+		meddbg("Unsupported internet media type: %s\n", mContentType.c_str());
+	}
+
+	switch (audioType) {
+	case AUDIO_TYPE_MP3:
+	case AUDIO_TYPE_AAC: {
+		// wait for audio stream data
+		if (!mCondv.wait_for(lock, WAIT_DATA_TIMEOUT, [=]{ return mIsDataReceived; })) {
```

Comment:
I think it's not until I get the data once,
You have to wait until the buffer is threshhold.

Reply:
yes, that's more reasonable, I will fix it.

Reply:
Fixed, refer to `void HttpInputDataSource::onBufferUpdated(ssize_t change, size_t current)` at line 242.